packages/oauth-helper/src/index.ts

```diff
@@ -0,0 +1,468 @@
+
+    // generate state
+    const state = await this.stateStore.generateStateParam(new Date(), options.metadata);
+    URL = `${URL}&state=${state}`;
+
+    // client id
+    URL = `${URL}&client_id=${this.clientId}`;
+
+    // redirect uri
+    if (options.redirectUri != null) {
+      URL = `${URL}&redirect_uri=${options.redirectUri}`;
+    }
+
+    // team id
+    if (options.teamId != null) {
+      URL = `${URL}&team=${options.teamId}`;
+    }
+
+    // user scope, only available for OAuth v2
+    if (options.userScopes != null && this.authVersion === 'v2') {
+      let userScopes;
+      if (options.userScopes instanceof Array) {
+        userScopes = options.userScopes.join(',');
+      } else {
+        userScopes = options.userScopes;
+      }
+      URL = `${URL}&user_scope=${userScopes}`;
+    }
+
+    return new Promise<string>((resolve) => {
+      resolve(URL);
+    });
+  }
+
+  /**
+   * This method handles the incoming request to the callback URL.
+   * It can be used as a RequestListener in almost any HTTP server
+   * framework.
+   *
+   * Verifies the state using the stateStore, exchanges the grant in the
+   * query params for an access token, and stores token and associated data
+   * in the installationStore.
+   */
+  public async handleCallback(
+    req: IncomingMessage,
+    res: ServerResponse,
+    options?: CallbackOptions,
+  ): Promise<void> {
+    console.log('handleCallback');
+    console.log(options);
+    let parsedUrl;
+    let code: string = '';
+    let state: string = '';
+
+    if (req.url != null) {
+      parsedUrl = url.parse(req.url, true);
+      code = parsedUrl.query.code as string;
+      state = parsedUrl.query.state as string;
+    }
+    console.log('code', code);
+    console.log('state', state);
+
+    try {
+      const metadata = await this.stateStore.verifyStateParam(new Date(), state);
+      console.log(metadata);
+
+      const client = new WebClient('');
+
+      let resp;
+      let installation: Installation;
+      console.log('authVersion', this.authVersion);
+      if (this.authVersion === 'v1') {
+        // convert response type from WebApiCallResult to OAuthResponse
+        resp = await client.oauth.access({
+          code,
+          client_id: this.clientId,
+          client_secret: this.clientSecret,
+        }) as unknown as OAuthV1Response;
+
+        // TODO: need to get appID
+        // resp obj for v1 - https://api.slack.com/methods/oauth.access#response
+        // TODO: look into getting bot.id added to platform response
+        // check if it already exists for v1
+        installation = {
+          team: { id: resp.team_id, name: resp.team_name },
+          appId: undefined, // not included in v1 unless workspace apps, so most likely undefinied
+          user: {
+            token: resp.access_token,
+            scopes: resp.scope.split(','),
+            id: '',
+          },
+          bot: {
+            // TODO is this correct?
```

Comment:
I'm just copying the top level scope into the bot scopes field here. Doesn't feel right.

Reply:
I don't actually know what to put in here? Is there some data we get during `oauth.access` that makes sense? Look at https://api.slack.com/methods/oauth.access#response, I only see a top level `scope` argument in the payload. Does it make sense to put in `user.scopes` or `bot.scopes`? 

For `oauth.v2.access`, the top level `scope` argument in the payload goes into `bot.scopes`, and `user.scopes` get `resp.authed_user.scope != null ? resp.authed_user.scope.split(',') : resp.authed_user.scope`

Reply:
You're right. 

>Does it make sense to put in user.scopes or bot.scopes?

It's not ideal but it makes sense to me, too.